generate_directory_indexes.py

```diff
@@ -115,47 +159,133 @@ def format_directory_listing(directory_listing):
+        full_path = os.path.join(current_path, file_name)
+        if is_excluded_file(file_name):
             continue
-        if(os.path.exists(fullpath)):
-            icon = 'folder.gif' if os.path.isdir(fullpath) else 'unknown.gif'
-            directory_listing.append({
-                'name': file,
-                'lastModified': os.path.getmtime(fullpath),
-                'size': os.path.getsize(fullpath),
-                'icon': icon
-            })
+        if os.path.exists(full_path):
+            icon = 'folder.gif' if os.path.isdir(full_path) else 'unknown.gif'
+            file_details.append(file_information(full_path, file_name))
         else:
-            logging.error('skipping \'{}\' because the file cannot be read'.format(fullpath))
+            logging.error('skipping \'{}\' because the file cannot be read'.format(full_path))
+
+    return file_details
+
+
+def make_index_files(base_path, current_path, file_details):
     for order_by in ['name', 'size', 'lastModified']:
         for reverse_order in [True, False]:
-            file_name = os.path.join(path, index_file_name('', order_by, reverse_order))
-            rendered_html = render_index(path, order_by, directory_listing, reverse_order, base_path)
-            if args.noop:
-                logging.info('Would create: ', file_name)
+            file_name = os.path.join(current_path, index_file_name('', order_by, reverse_order))
+            rendered_html = render_index(current_path, order_by, file_details,
+                                         reverse_order, base_path)
+            if configuration.noop:
+                logging.info('Would create: {}'.format(file_name))
             else:
                 logging.info('Wrote: {}'.format(file_name))
                 index_file = open(file_name, 'w')
                 index_file.write(rendered_html)
                 index_file.close()
 
 
-    # recurse into subdirectories
-    for file in contents:
-        file_pwd = os.path.join(path, file)
-        if os.path.isdir(file_pwd):
-            walk(file_pwd, base_path)
+def traverse_tree(base_path, current_path, file_metadata = None):
+    contents = os.listdir(current_path)
+
+    if file_metadata != None:
+        file_details = parse_file_metadata(current_path, file_metadata)
+    else:
+        file_details = gather_file_details(current_path, contents)
+
+    make_index_files(base_path, current_path, file_details)
+    for file_name in contents:
+        absolute_path = os.path.join(current_path, file_name)
+        if os.path.isdir(absolute_path):
+            traverse_tree(base_path, absolute_path, file_metadata)
+
+
+def validate_input(configuration):
+    global script
+
+    if not os.path.isdir(configuration.path):
+        sys.exit('{}: ERROR {} is not a directory'.format(script, configuration.path))
+
 
-def validate_input(path):
-    if not os.path.isdir(path):
-        sys.exit('Error: {} is not a directory'.format(path))
+parse_arguments()
```

Comment:
We should push these methods into some sort of boilerplate for python so that you can easily debug the file in a repl.

Try this instead:
```py
if __name__ == '__main__':
    parse_arguments()
    validate_input(configuration)
    traverse_tree(configuration.path, configuration.path, configuration.file_metadata)
```